modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -970,28 +976,41 @@ private CompletableFuture<Void> continueIndexScan(
                         return completedFuture(null); // End of range reached. Exit loop.
                     }
 
-                    return lockManager.acquire(txId, new LockKey(tableId(), currentRow.rowId()), LockMode.S)
+                    RowId rowId = currentRow.rowId();
+
+                    return lockManager.acquire(txId, new LockKey(tableId(), rowId), LockMode.S)
                             .thenComposeAsync(rowLock -> { // Table row S lock
-                                ReadResult readResult = mvDataStorage.read(currentRow.rowId(), HybridTimestamp.MAX_VALUE);
-                                return resolveAndCheckReadCompatibility(readResult, txId)
-                                        .thenCompose(resolvedReadResult -> {
-                                            if (resolvedReadResult != null) {
-                                                if (indexRowMatches(currentRow, resolvedReadResult, schemaAwareIndexStorage)) {
-                                                    result.add(resolvedReadResult);
-                                                }
-                                            }
-
-                                            // Proceed scan.
-                                            return continueIndexScan(
-                                                    txId,
-                                                    schemaAwareIndexStorage,
-                                                    indexLocker,
-                                                    indexCursor,
-                                                    batchSize,
-                                                    result,
-                                                    isUpperBoundAchieved
-                                            );
-                                        });
+                                ReadResult readResult = mvDataStorage.read(rowId, HybridTimestamp.MAX_VALUE);
+                                return resolveAndCheckReadCompatibility(readResult, txId, () -> {
+                                    if (readResult.newestCommitTimestamp() == null) {
```

Comment:
Same as above. I'd rather hide given logic inside resolve... method if it's possible.

Reply:
The logic is already written and used in other places. In cases like the above, we have to get the commit timestamp earlier than the resolution is finalized.